File: lib/reporter.rb
# MP3 consistency reporter
#
# Author: Falko Schmidt <kaethorn@gmail.com>

require 'find'
require 'mimemagic'
require 'haml'
require_relative 'scanner_mp3'
require_relative 'scanner_ogg_vorbis'
require_relative 'scanner_flac'

class Reporter

  def initialize(directory, report_type)
    @directory = directory
    @report_type = report_type
    @report = {}
  end

  def run
    iterate_directory
    generate
  end

  private

  # Generates a report for the specified directory
  def iterate_directory
    Find.find(@directory) do |file|
      # Only consider files
      unless FileTest.directory?(file)
        raw_file = File.open file
        fileType = MimeMagic.by_path(file)
        magicType = MimeMagic.by_magic(raw_file)
        raw_file.close

        # Only consider audio files
        if fileType == 'audio/mpeg' and magicType == 'audio/mpeg'
          mp3_scanner = MP3Scanner.new file, @report
          mp3_scanner.scan
        elsif fileType == 'audio/ogg' and magicType == 'audio/x-vorbis+ogg'
          ogg_vorbis_scanner = OggVorbisScanner.new file, @report
          ogg_vorbis_scanner.scan
        elsif fileType == 'audio/flac' and magicType == 'audio/flac'
          flac_scanner = FLACScanner.new file, @report
          flac_scanner.scan
        elsif fileType == 'audio/mp4' and magicType == 'audio/mp4'
          # MP4 files are not yet supported in taglib-ruby
        end
      end
    end
  end

  def generate
    template = File.read("report.#{@report_type}.haml")
    haml_engine = Haml::Engine.new(template)
    output = File.new 'report.html', 'w'
    output.write haml_engine.render(Object.new, :report => @report)
    output.close
  end
end

File: lib/reporter.cxx
#include "reporter.h"
#include "boost/filesystem.hpp"
#include <magic.h>
#include "scanner_mp3.h"

namespace fs = boost::filesystem;

Reporter::Reporter(string dir, string r_type) {
  directory = dir;
  report_type = r_type;
  report = map< string, map< string, map< string, vector<string> > > >();
}

void Reporter::run() {
  iterate_directory();
  generate();
}

void Reporter::generate() {
  /*
    template = File.read("report.#{@report_type}.haml")
    haml_engine = Haml::Engine.new(template)
    output = File.new 'report.html', 'w'
    output.write haml_engine.render(Object.new, :report => @report)
    output.close
  */
}

const string Reporter::get_file_type(const string file) {
  const char *contentType; 
  string type;

  magic_t magic = magic_open(MAGIC_MIME_TYPE|MAGIC_ERROR); 
  if (magic == NULL) { 
    cerr << file << ": Error initializing libmagic" << endl;
    return "unknown";
  } 
  if (magic_load(magic, NULL) != 0) {
    cerr << file << ": Error loading libmagic database" << endl;
    magic_close(magic);
    return "unknown";
  }
  contentType = magic_file(magic, file.c_str()); 
  if (contentType == NULL) { 
    cerr << file << ": Unknown content type" << endl; 
    magic_close(magic);
    return "unknown";
  } 
  type = string(contentType);
  magic_close(magic);
  return type;
}

void Reporter::iterate_directory() {
  for (fs::recursive_directory_iterator end, file(directory); 
      file != end; ++file) {
    if (is_directory(file->status())) {
      continue;
    }
    string file_type = get_file_type(file->path().string());
    // Only consider audio files
    if (file_type == "audio/mpeg") {
      MP3Scanner mp3_scanner(file->path().string(),&report);
      mp3_scanner.scan();
    } else if (file_type == "application/ogg") {
    } else if (file_type == "audio/x-flac") {
    } else if (file_type == "audio/mp4") {
    }
  }
}
